api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/observer/MetricObserver.java

```diff
@@ -0,0 +1,142 @@
+        jsonOutputStream = new ByteArrayOutputStream();
+        try {
+            generator = mapper.getFactory().createGenerator(response.getOutputStream(), JsonEncoding.UTF8);
+            generator.writeStartArray();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void onNext(Metric<T> metric) {
+        try {
+            if (count.incrementAndGet() == 1) {
+                response.setStatus(HttpServletResponse.SC_OK);
+                response.setHeader("Content-Type", "application/json");
+            }
+            generator.writeStartObject();
+            generator.writeStringField("id", metric.getMetricId().getName());
+            generator.writeStringField("tenantId", metric.getMetricId().getTenantId());
+            generator.writeStringField("type", metric.getMetricId().getType().toString());
+            if (!metric.getTags().isEmpty()) {
+                generator.writeObjectFieldStart("tags");
+                for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
+                    generator.writeStringField(tag.getKey(), tag.getValue());
+                }
+                generator.writeEndObject();
+            }
+            if (metric.getDataRetention() != null) {
+                generator.writeNumberField("dataRetention", metric.getDataRetention());
+            }
+            if (metric.getMinTimestamp() != null) {
+                generator.writeNumberField("minTimestamp", metric.getMinTimestamp());
+            }
+            if (metric.getMaxTimestamp() != null) {
+                generator.writeNumberField("maxTimestamp", metric.getMaxTimestamp());
+            }
+            generator.writeEndObject();
+            generator.flush();
+        } catch (IOException e) {
+            throw new RuntimeException("Streaming data to client failed", e);
+        }
+    }
+
+    @Override
+    public void onError(Throwable e) {
+        logger.warn("Fetching data failed", e);
+        try {
+            if (count.get() == 0) {
+                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+                response.setHeader("Content-Type", "application/json");
+                ApiError apiError = new ApiError(Throwables.getRootCause(e).getMessage());
+                mapper.writeValue(response.getOutputStream(), apiError);
+            } else {
+                generator.close();
+            }
+        } catch (IOException ie) {
+            logger.warn("There was an error closing the JSON generator", ie);
+        } finally {
+            request.getAsyncContext().complete();
+        }
+    }
+
+    @Override
+    public void onCompleted() {
+        try {
+            if (count.get() == 0) {
+                response.setStatus(HttpServletResponse.SC_NO_CONTENT);
+            } else {
+                generator.writeEndArray();
+            }
+            generator.close();
```

Comment:
This is never called if we encounter an error (onError terminates the stream, so onCompleted is never called)